File: src/main/java/me/testaccount666/serversystem/commands/executables/moderation/TabCompleterModeration.java
package me.testaccount666.serversystem.commands.executables.moderation;

import com.destroystokyo.paper.profile.PlayerProfile;
import io.papermc.paper.ban.BanListType;
import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.commands.interfaces.ServerSystemTabCompleter;
import me.testaccount666.serversystem.managers.PermissionManager;
import me.testaccount666.serversystem.userdata.CachedUser;
import me.testaccount666.serversystem.userdata.OfflineUser;
import me.testaccount666.serversystem.userdata.User;
import me.testaccount666.serversystem.userdata.UserManager;
import org.bukkit.BanEntry;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;
import org.bukkit.command.Command;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class TabCompleterModeration implements ServerSystemTabCompleter {

    private static final List<String> _TIME_UNITS = List.of("s", "m", "h", "d", "w", "mo", "y");
    private static final List<String> _NUMBERS = IntStream.range(0, 10).mapToObj(String::valueOf).toList();

    private static final String _PERMANENT = "permanent";

    @Override
    public Optional<List<String>> tabComplete(User commandSender, Command command, String label, String... arguments) {
        var commandName = command.getName().toLowerCase();
        if (!hasPermission(commandSender, commandName)) return Optional.of(Collections.emptyList());

        return switch (arguments.length) {
            case 1 -> handlePlayerNameCompletion(commandName, arguments[0]);
            case 2 -> handleTimeCompletion(arguments[1], commandName);
            default -> Optional.empty();
        };
    }

    private boolean hasPermission(User commandSender, String commandName) {
        var permissionNode = switch (commandName) {
            case "mute" -> "Moderation.Mute.Use";
            case "unmute" -> "Moderation.Mute.Remove";
            case "shadowmute" -> "Moderation.Mute.Shadow";
            case "ban" -> "Moderation.Ban.Use";
            case "unban" -> "Moderation.Ban.Remove";
            default -> null;
        };
        return PermissionManager.hasCommandPermission(commandSender, permissionNode, false);
    }

    private Optional<List<String>> handlePlayerNameCompletion(String commandName, String argument) {
        var possibleCompletions = getPlayerNames(commandName);

        if (argument.isEmpty()) return Optional.of(possibleCompletions);

        var filteredCompletions = possibleCompletions.stream()
                .filter(name -> name.toLowerCase().startsWith(argument.toLowerCase()))
                .toList();

        return Optional.of(filteredCompletions);
    }

    private List<String> getPlayerNames(String commandName) {
        if (commandName.equalsIgnoreCase("unban")) return Bukkit.getServer().getBanList(BanListType.PROFILE).getEntries().stream()
                .map(BanEntry::getBanTarget)
                .map(PlayerProfile.class::cast)
                .map(PlayerProfile::getName).filter(Objects::nonNull).toList();

        if (commandName.equalsIgnoreCase("unmute")) return ServerSystem.Instance.getRegistry().getService(UserManager.class)
                .getCachedUsers().stream()
                .map(CachedUser::getOfflineUser)
                .filter(user -> user.getMuteManager().hasActiveModeration()).map(OfflineUser::getName)
                .filter(Optional::isPresent).map(Optional::get).toList();

        return Arrays.stream(Bukkit.getOfflinePlayers())
                .map(OfflinePlayer::getName)
                .filter(Objects::nonNull).toList();
    }

    private Optional<List<String>> handleTimeCompletion(String argument, String commandName) {
        if (isRemoveCommand(commandName)) return Optional.empty();
        if (argument.isEmpty()) return Optional.of(getDefaultTimeSuggestions());

        var suggestions = new ArrayList<String>();

        if (_PERMANENT.startsWith(argument.toLowerCase())) suggestions.add(_PERMANENT);

        if (argument.matches(".*\\d$")) suggestions.addAll(getTimeSuggestions(argument));
        else if (argument.matches("\\d+")) suggestions.addAll(getTimeSuggestions(argument));
        else suggestions.addAll(getMatchingTimeUnits(argument));

        return Optional.of(suggestions);
    }

    private List<String> getDefaultTimeSuggestions() {
        var suggestions = new ArrayList<>(_NUMBERS);
        suggestions.add(_PERMANENT);
        return suggestions;
    }

    private List<String> getTimeSuggestions(String argument) {
        var suggestions = new ArrayList<String>();
        _NUMBERS.forEach(num -> suggestions.add(argument + num));
        _TIME_UNITS.forEach(unit -> suggestions.add(argument + unit));
        return suggestions;
    }

    private List<String> getMatchingTimeUnits(String argument) {
        return _TIME_UNITS.stream()
                .filter(unit -> unit.startsWith(argument.toLowerCase()))
                .collect(Collectors.toList());
    }

    private boolean isRemoveCommand(String commandName) {
        return commandName.toLowerCase().startsWith("un");
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/moderation/TabCompleterModeration.kt
package me.testaccount666.serversystem.commands.executables.moderation

import com.destroystokyo.paper.profile.PlayerProfile
import io.papermc.paper.ban.BanListType
import me.testaccount666.serversystem.commands.interfaces.ServerSystemTabCompleter
import me.testaccount666.serversystem.commands.interfaces.getService
import me.testaccount666.serversystem.managers.PermissionManager.hasCommandPermission
import me.testaccount666.serversystem.userdata.CachedUser
import me.testaccount666.serversystem.userdata.User
import me.testaccount666.serversystem.userdata.UserManager
import org.bukkit.BanEntry
import org.bukkit.Bukkit
import org.bukkit.command.Command

class TabCompleterModeration : ServerSystemTabCompleter {
    override fun tabComplete(commandSender: User, command: Command, label: String, vararg arguments: String): List<String>? {
        val commandName = command.name.lowercase()
        if (!hasPermission(commandSender, commandName)) return listOf()

        return when (arguments.size) {
            1 -> handlePlayerNameCompletion(commandName, arguments[0])
            2 -> handleTimeCompletion(arguments[1], commandName)
            else -> null
        }
    }

    private fun hasPermission(commandSender: User, commandName: String): Boolean {
        val permissionNode = when (commandName) {
            "mute" -> "Moderation.Mute.Use"
            "unmute" -> "Moderation.Mute.Remove"
            "shadowmute" -> "Moderation.Mute.Shadow"
            "ban" -> "Moderation.Ban.Use"
            "unban" -> "Moderation.Ban.Remove"
            else -> null
        }
        return hasCommandPermission(commandSender, permissionNode, false)
    }

    private fun handlePlayerNameCompletion(commandName: String, argument: String): List<String> {
        val possibleCompletions = getPlayerNames(commandName)
        if (argument.isEmpty()) return possibleCompletions

        return possibleCompletions.filter { it.startsWith(argument, true) }
    }

    private fun getPlayerNames(commandName: String): List<String> {
        if (commandName.equals("unban", true)) {
            return Bukkit.getServer().getBanList(BanListType.PROFILE)
                .getEntries<BanEntry<PlayerProfile>>()
                .map { it.getBanTarget() }.mapNotNull { it.name }
        }

        if (commandName.equals("unmute", true)) {
            return getService<UserManager>()
                .cachedUsers.asSequence()
                .map(CachedUser::offlineUser).filter { it.muteManager.hasActiveModeration() }
                .mapNotNull { it.getNameOrNull() }.toList()
        }

        return Bukkit.getOfflinePlayers().mapNotNull { it.name }
    }

    private fun handleTimeCompletion(argument: String, commandName: String): List<String>? {
        if (isRemoveCommand(commandName)) return null
        if (argument.isEmpty()) return defaultTimeSuggestions

        val suggestions = mutableListOf<String>()

        if (_PERMANENT.startsWith(argument, true)) suggestions.add(_PERMANENT)

        if (argument.matches(".*\\d$".toRegex()) || argument.matches("\\d+".toRegex())) {
            suggestions.addAll(getTimeSuggestions(argument))
        } else {
            suggestions.addAll(getMatchingTimeUnits(argument))
        }

        return suggestions
    }

    private val defaultTimeSuggestions by lazy { _NUMBERS + _PERMANENT }

    private fun getTimeSuggestions(argument: String) = _NUMBERS.map { argument + it } + _TIME_UNITS.map { argument + it }

    private fun getMatchingTimeUnits(argument: String) = _TIME_UNITS.filter { it.startsWith(argument, true) }

    private fun isRemoveCommand(commandName: String) = commandName.startsWith("un", true)

    companion object {
        private val _TIME_UNITS = listOf("s", "m", "h", "d", "w", "mo", "y")
        private val _NUMBERS = (0 until 10).map { it.toString() }

        private const val _PERMANENT = "permanent"
    }
}
